aten/src/ATen/TensorNames.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include <ATen/core/EnableNamedTensor.h>
+#include <ATen/WrapDimUtils.h>
+
+namespace at { namespace namedinference {
+
+#ifdef BUILD_NAMEDTENSOR
+
+// TensorName and TensorNames are wrappers around Dimname and DimnameList
+// that contain helper functions to make writing name inference rules easier.
+//
+// A TensorName represents a Dimname associated with some DimnameList (from a Tensor).
+// This encapsulates all the information that is needed to check if names *match*
+// and to *unify* names.
+//
+// Definition: Two names in two tensors *match* if they are equal, or if at
+// least one of them is a wildcard that can be *refined* to the other name.
+//
+// Here is an example of checking if two names match.
+// tensor: Tensor[A, None]
+// other: Tensor[A]
+//
+// Let's say we wish to check if tensor.names[-1] matches other.names[-1].
+// None (in tensor) cannot match A (in other) because if the None were refined
+// to A, `tensor` would have duplicate names [A, A]. Therefore we need to check
+// tensor.names [A, None] for the existence of A.
+struct CAFFE2_API TensorName {
+  explicit TensorName(ArrayRef<Dimname> origin, int origin_idx)
+    : origin_(origin),
+      origin_idx_(origin_idx),
+      name_(origin[maybe_wrap_dim(origin_idx, origin.size())]) {}
```

Comment:
it's not directly related to this PR but at some point we should add better comments for maybe_wrap_dim.

Reply:
What sort of comments did you want to see for maybe_wrap_dim? Just describing what it does?

Reply:
right, I think I've re-read that code a few times come across the function. When I see "wrap" I think composition and get confused so go check the definition, but there's no comment and the code in WrapDimMinimal.h takes a bit of thinking to understand.

Reply:
> When I see "wrap" I think composition

Out of curiosity, what do you mean by this? Something like monads?

Reply:
I just mean when you 'wrap' some object in another object (making it a member) as in:
https://en.wikipedia.org/wiki/Adapter_pattern
https://en.wikipedia.org/wiki/Composition_over_inheritance
Or when we wrap some function call by calling it from some other function:
https://en.wikipedia.org/wiki/Wrapper_function

In some sense this particular PR can be thought of as wrapping dimname/dimnamelist instances in a class that adds extra functionality, which I think contributes to me not immediately jumping to the other meaning of index wrapping when I see this function call.

Reply:
Actually you refer to exactly what I meant, above:
// TensorName and TensorNames are wrappers around Dimname and DimnameList
that meaning being different from the `maybe_wrap_dim` meaning of 'wrap'.